### DISCUSSION TITLE
[SUGGESTION] Don't accept implicit integers in [[assert: 1]]

### DISCUSSION THREAD

**Discussion by realgdman** [hsutter/cppfront#645] [Suggestions]
This is minor one.
When skimming through some snippets, I got impression that you can write
`[[assert: 42]]`
Which is actually not checking for value, (say, of short-syntax function), and just implicit converted to `bool`

According to [this note at ES.87](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#note-226), checking for pointers is good, but checking integers without comparison can be suspicious.

Probably assert accepting only bools or pointers can be implemented with the help of concepts like std::is_arithmetic_v and std::is_pointer_v instead of simply `bool`
https://github.com/hsutter/cppfront/blob/f83ca9b727e9c07f60a1933584a49a3d971936f5/include/cpp2util.h#L325

**Comment by JohelEGP** [comment]
**We should still accept function pointers.**

This is also valid: `f: () [[pre: :() = 1]] = { }`, https://godbolt.org/z/a7db78azx.

<details><summary>Generated C++1.</summary>

```C++
//=== Cpp2 type declarations ====================================================


#include "cpp2util.h"


//=== Cpp2 type definitions and function declarations ===========================

#line 1 "/app/main.cpp2"
auto f() -> void;
auto main() -> int;

//=== Cpp2 function definitions =================================================

#line 1 "/app/main.cpp2"
auto f() -> void{
                           cpp2::Default.expects([]() -> void { 1; }, "");
#line 1 "/app/main.cpp2"
}
auto main() -> int{}
```

</details>
